### PR TITLE
tls: Fix tls.Server constructor and TlsOptions

### PR DESCRIPTION
**Pull request by dylan-hart**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

## Context for the suggested changes:

Documentation for tls.createServer() for Node.js 14.x, 15.x (and others) states that any net.createServer options can be provided in the `options` object parameter ([see here](https://nodejs.org/docs/latest-v15.x/api/tls.html#tls_tls_createserver_options_secureconnectionlistener)). This is seen to be true in the Node.js 15.x source [here](https://github.com/nodejs/node/blob/v15.x/lib/_tls_wrap.js#L1222) because the same parameter is passed unchanged to the net.Server constructor.

However, the current definition of the parameter's type, `TlsOptions` ([seen here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/tls.d.ts#L388)) does not define these net.createServer / net.Server options. This doesn't commonly cause linter/compiler issues because all of the net.createServer / net.Server options are [optional](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/net.d.ts#L259). The linter will have a problem, though, if the options object is explicitly declared using the `TlsOptions` type:

```typescript
import { TlsOptions, Server } from 'tls'

const options: TlsOptions = {};
const server = new Server(options, () => {}); // Problem! Type 'TlsOptions' has no properties in common with type '{ allowHalfOpen?: boolean; pauseOnConnect?: boolean }'.
```

Additionally, direct use of the net.Server constructor is typed [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/net.d.ts#L184), but in tls.Server is not updated to use `TlsOptions`.

This PR adds the missing net.Server options to the `TlsOptions` interface and re-declares constructors inherited from net.Server using the `TlsOptions` interface. A test demonstrating the linter error prior to these changes is also included.

Note: I would have preferred to extend a declared interface of the net.Server options, but no explicit interface declaration exists and creating one now would affect more projects than this PR needs to. Duplicating the options within the `TlsOptions` interface seems preferable.

### CODE REVIEW

**Comment by dylan-hart** [comment]
The errors logged in CI / test (pull_request) also occur in a fresh checkout of https://github.com/DefinitelyTyped/DefinitelyTyped master. Should I be concerned with this?

**Comment by Flarna** [comment]
@dynlan-hart Yes, master is broken since a while. Can be seen also there https://github.com/DefinitelyTyped/DefinitelyTyped/#current-status

Root cause is some issue in typescript 4.2, some other node PR here linked to it but I can't remember which one.

There is not much you can do regarding this. Problem is that PRs with failing CI are not considered for merging by the maintainers. So be prepared that this one is stalled for a while. I know this is not nice but I can't provide any better news.

File: types/node/tls.d.ts [hunk 1]
```diff
@@ -432,6 +432,16 @@ declare module "tls" {
          * emitted with `ERR_TLS_PSK_SET_IDENTIY_HINT_FAILED` code.
          */
         pskIdentityHint?: string;
+
+        /**
+         * Indicates whether half-opened TCP connections are allowed. __Default:__ `false`.
+         */
+        allowHalfOpen?: boolean;
```
Comment (Flarna) [inline types/node/tls.d.ts:439]: What about extracting the options for net server into e.g. `NetServerOptions` and add it to the extends list in `TlsOptions` ?

**Comment by dylan-hart** [comment]
> @dynlan-hart Yes, master is broken since a while. Can be seen also there https://github.com/DefinitelyTyped/DefinitelyTyped/#current-status
> 
> Root cause is some issue in typescript 4.2, some other node PR here linked to it but I can't remember which one.
> 
> There is not much you can do regarding this. Problem is that PRs with failing CI are not considered for merging by the maintainers. So be prepared that this one is stalled for a while. I know this is not nice but I can't provide any better news.

Ah ok. Thanks. I suspected but this being my first time here - lots to look at ... not entirely sure it wasn't something I was missing.

File: types/node/tls.d.ts (hunk 1 above)
Comment (dylan-hart) [reply types/node/tls.d.ts:439]: I'd prefer something like this. I was concerned that editing "net" definitions might be considered changing a critical package unnecessarily, but I'll be easily persuaded it's the better option.

If `NetServerOptions` sounds like a good idea, I'll make the changes and modify this PR next chance I get.

Comment (Flarna) [reply types/node/tls.d.ts:439]: There are already two places in net (`class Server` and `createServer`) where this new type can be used.

Looking once more into the net.d.ts file it seems that there `XXXOpts` is usually used so maybe go for `ServerOpts`. Don't think the prefix `Net` is needed as it is anyway namespaced within the net module.

Comment (dylan-hart) [reply types/node/tls.d.ts]: Agree - I see nine types in net.d.ts using `XXXOpts`, and just one using `XXXOptions`.

**Review by Flarna — APPROVED** [review]

**Comment by Flarna** [comment]
Regarding the CI fail: Once https://github.com/microsoft/TypeScript/pull/42447 is resolved it should go away - and hopefully no new issue slips is...
You should merge upstream/master once this happened (or rebase). This re-triggers CI.

**Comment by dylan-hart** [comment]
> Regarding the CI fail: Once [microsoft/TypeScript#42447](https://github.com/microsoft/TypeScript/pull/42447) is resolved it should go away - and hopefully no new issue slips is...
> You should merge upstream/master once this happened (or rebase). This re-triggers CI.

Thanks! I'll keep an eye out.

**Review by Flarna — APPROVED** [review]

**Review by a-tarasyuk — APPROVED** [review]

**Review by amcasey — APPROVED** [review]
Thanks!  Sorry for the CI pain.

**Comment by dylan-hart** [comment]
Ready to merge

**Pull request merged**